File: backend/app/routers/research.py
"""Research endpoint: take a topic, return perspective-tagged web sources.

Thin HTTP wrapper around `app.pipeline.research.research_topic`. The pipeline call
is synchronous and takes ~10-20s (it plans queries with an LLM, then hits Tavily);
defining the handler with `def` (not `async def`) lets FastAPI run it in a threadpool
so the event loop isn't blocked.
"""

from fastapi import APIRouter
from pydantic import BaseModel

from app.pipeline.research import research_topic
from app.schemas.source import SourceItem

router = APIRouter(prefix="/api", tags=["research"])


class ResearchRequest(BaseModel):
    topic: str


class ResearchResponse(BaseModel):
    topic: str
    sources: list[SourceItem]


@router.post("/research", response_model=ResearchResponse)
def research(req: ResearchRequest) -> ResearchResponse:
    """Plan angled searches for `topic` and return deduped, tagged sources."""
    sources = research_topic(req.topic)
    return ResearchResponse(topic=req.topic, sources=sources)

File: frontend/src/lib/research.ts
import { apiFetch } from '@/lib/api'

// Mirrors backend app/schemas/source.py : SourceItem
export type Perspective = 'mainstream' | 'anti_imperialist'

export interface SourceItem {
  title: string
  url: string
  snippet: string
  perspective: Perspective
}

// Mirrors backend app/routers/research.py : ResearchResponse
export interface ResearchResponse {
  topic: string
  sources: SourceItem[]
}

export function postResearch(topic: string): Promise<ResearchResponse> {
  return apiFetch<ResearchResponse>('/api/research', {
    method: 'POST',
    body: JSON.stringify({ topic }),
  })
}
